### PR TITLE
Updatable owners

### PR DESCRIPTION
**Pull request by krebernisak**

### CODE REVIEW

File: contracts/contracts/multisig/Multisig.cairo [hunk 1]
```diff
@@ -121,11 +132,9 @@ func require_not_executed{
 end
 
 # Revert if tx has been confirmed for the calling account already
-func require_not_confirmed{
-        syscall_ptr : felt*,
-        pedersen_ptr : HashBuiltin*,
-        range_check_ptr
-    }(tx_index : felt):
+func require_not_confirmed{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
```
Comment (krebernisak) [inline contracts/contracts/multisig/Multisig.cairo]: Did you manually change formatting or did you use a specific formatter? 
If you used a formatter, which one?

File: contracts/contracts/multisig/Multisig.cairo [hunk 2]
```diff
@@ -148,41 +155,40 @@ func require_confirmed{
     return ()
 end
 
+func require_valid_tx{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(tx_index):
+    let (tx_valid_since) = _tx_valid_since.read()
+
+    with_attr error_message("invalid transaction index"):
```
Comment (krebernisak) [inline contracts/contracts/multisig/Multisig.cairo]: Consider adding a more descriptive error here:

```suggestion
    with_attr error_message("tx invalidated: config changed after submission"):
```

File: contracts/contracts/multisig/Multisig.cairo [hunk 3]
```diff
@@ -326,52 +308,45 @@ end
...
-        confirmations_required : felt,
-    ):
+func constructor{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    owners_len : felt, owners : felt*, confirmations_required : felt
+):
+    const lower_bound = 1
+
     with_attr error_message("invalid number of required confirmations"):
-        assert_le(confirmations_required, owners_len)
+        assert_in_range(confirmations_required, lower_bound, owners_len + 1)
```
Comment (krebernisak) [inline contracts/contracts/multisig/Multisig.cairo]: Here we force at least 1 confirmation, but we don't force at least 1 owner. This means the ownership of the multisig can be burned (set to 0 owners), which could be a desired features for some use cases.

Thinking out loud if we are ok with the design 🤔  cc @archseer

File: contracts/contracts/multisig/Multisig.cairo [hunk 4]
```diff
@@ -326,52 +308,45 @@ end
...
+    end
+
+    with_attr error_message("owners has to be unique"):
```
Comment (krebernisak) [inline contracts/contracts/multisig/Multisig.cairo]: nit: Consider making this message consistent with other error messages, where we don't point to what needs to be but what is invalid:

```suggestion
    with_attr error_message("owners not unique"):
```

File: contracts/contracts/multisig/Multisig.cairo [hunk 5]
```diff
@@ -383,23 +358,21 @@ func submit_transaction{
...
-        range_check_ptr
-    }(tx_index : felt):
+func confirm_transaction{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    tx_index : felt
+):
     require_owner()
     require_tx_exists(tx_index=tx_index)
     require_not_executed(tx_index=tx_index)
+    require_valid_tx(tx_index=tx_index)
+    # TODO: clarify if needed
```
Comment (krebernisak) [inline contracts/contracts/multisig/Multisig.cairo]: I'd suggest renaming as `require_tx_valid`, switching order (both to be more consistent and readable), then add necessary inline comments to the `require_tx_valid` fn about what constitutes a valid tx.

```suggestion
    require_tx_exists(tx_index=tx_index)
    require_tx_valid(tx_index=tx_index)
    require_not_executed(tx_index=tx_index)
```

File: contracts/contracts/multisig/Multisig.cairo [hunk 6]
```diff
@@ -148,41 +155,40 @@ func require_confirmed{
     return ()
 end
 
+func require_valid_tx{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(tx_index):
```
Comment (krebernisak) [inline contracts/contracts/multisig/Multisig.cairo]: Please add docs here (fn) about what constitutes an in/valid tx.

File: contracts/contracts/multisig/Multisig.cairo [hunk 7]
```diff
@@ -474,24 +436,113 @@ func execute_transaction{
...
+@external
+func set_confirmations_required{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    confirmations_required : felt
+):
+    const lower_bound = 1
+
+    let (caller) = get_caller_address()
+    let (contract_address) = get_contract_address()
+
+    with_attr error_message("caller shall be multisig"):
```
Comment (krebernisak) [inline contracts/contracts/multisig/Multisig.cairo]: Can we extract the `require_quorum` fn from here, and use it in both `set_confirmations_required` and `set_owners`?

Additionally, I've noticed the `require_owner` fn call is part of `submit_transaction` & `execute_transaction`. I think anyone should be able to execute the tx if M/N owners signed, and perhaps also anyone should be able to submit a transaction 🤔 

@archseer any suggestions on the above?

File: contracts/contracts/multisig/Multisig.cairo [hunk 8]
```diff
@@ -474,24 +436,113 @@ func execute_transaction{
...
+    end
+
+    let (owners_len) = _owners_len.read()
+
+    with_attr error_message("invalid number of required confirmations"):
+        assert_in_range(confirmations_required, lower_bound, owners_len + 1)
+    end
+
+    _confirmations_required.write(confirmations_required)
+    ConfirmationsSet.emit(confirmations_required)
```
Comment (krebernisak) [inline contracts/contracts/multisig/Multisig.cairo]: Also used in the constructor, where the event is not emitted.
Consider extracting as an internal fn and reuse.

File: contracts/contracts/multisig/Multisig.cairo [hunk 9]
```diff
@@ -474,24 +436,113 @@ func execute_transaction{
...
+    let (caller) = get_caller_address()
+    let (contract_address) = get_contract_address()
+
+    with_attr error_message("caller shall be multisig"):
+        assert caller = contract_address
+    end
+
+    with_attr error_message("invalid num of owners"):
+        assert_not_zero(owners_len)
+    end
```
Comment (krebernisak) [inline contracts/contracts/multisig/Multisig.cairo]: In the constructor we do currently allow this.

File: contracts/contracts/multisig/Multisig.cairo [hunk 10]
```diff
@@ -474,24 +436,113 @@ func execute_transaction{
...
+
+    # Clean previous owners
+    let (old_owners_len) = _owners_len.read()
+    _clean_owners(0, old_owners_len)
+
+    let (local confirmations_required) = _confirmations_required.read()
+    let (lt) = is_le(owners_len, confirmations_required - 1)
+    if lt == TRUE:
+        # Lower number of confirmation
+        _confirmations_required.write(owners_len)
```
Comment (krebernisak) [inline contracts/contracts/multisig/Multisig.cairo]: Event not emitted here, but they are when the `_confirmations_required` are changed directly. Consider using an internal fn for this logic and reuse so there are no gaps like this.

**Review by krebernisak — COMMENTED** [review]
Nice work @PacoEdwin! 

Here are some suggestions and open questions:

File: contracts/contracts/multisig/Multisig.cairo (hunk 1 above)
Comment (archseer) [reply contracts/contracts/multisig/Multisig.cairo]: There's a built-in `cairo-format`

File: contracts/contracts/multisig/util.cairo [hunk 11]
```diff
@@ -1,6 +1,46 @@
...
+from starkware.cairo.common.dict_access import DictAccess
+
+func almost_equal(a, b) -> (res : felt):
     if (a - b) * (a - b - 1) * (a - b + 1) == 0:
         return (res=1)
     end
     return (res=0)
 end
+
+func assert_unique_addresse_impl{range_check_ptr, dict_ptr : DictAccess*}(
```
Comment (archseer) [inline contracts/contracts/multisig/util.cairo]: adresses

File: contracts/contracts/multisig/Multisig.cairo (hunk 3 above)
Comment (archseer) [reply contracts/contracts/multisig/Multisig.cairo]: Serum multisig does checks the other way around: `owners_len` must be at least equal to threshold or more: https://github.com/project-serum/multisig/blob/d6fbe734dc4026b9c8fdd2f9e78904112150fa74/programs/multisig/src/lib.rs#L127-L129=

File: contracts/contracts/multisig/Multisig.cairo (hunk 7 above)
Comment (archseer) [reply contracts/contracts/multisig/Multisig.cairo]: Only one of the owners should be able to create a transaction:

https://github.com/project-serum/multisig/blob/d6fbe734dc4026b9c8fdd2f9e78904112150fa74/programs/multisig/src/lib.rs#L54-L55=

But I think it's fine to allow anyone to execute if we have quorum of confirmations.

File: contracts/contracts/multisig/Multisig.cairo (hunk 3 above)
Comment (taco-paco) [reply contracts/contracts/multisig/Multisig.cairo]: @krebernisak as @archseer pointed out all implementation I saw use such logic. Also its not clear for me what is the use case 0 owners and how at least 1 confirmation aligns with this? We want multisig to execute any transaction in that case? Looks like confirmations in that case pointless. 

That also would require a lot of change in the code

File: contracts/contracts/multisig/Multisig.cairo (hunk 7 above)
Comment (taco-paco) [reply contracts/contracts/multisig/Multisig.cairo]: > Can we extract the require_quorum fn from here, and use it in both set_confirmations_required and set_owners?

Sure, I would prefer `require_multisig/require_itself/require_caller_multisig` since here we're not checking if quorum requirements were met and we have enough confirmations.

> anyone should be able to execute the tx if M/N owners signed, and perhaps also anyone should be able to submit a transaction

I don't see anything bad in doing this but as @archseer pointed out serum have restriction on that only owners can submit but anyone can execute it. But this varies since gnosis for example checks for owner on execution as well https://github.com/gnosis/MultiSigWallet/blob/90639984c960d281bed3e0a5d56dd4adcb9407c4/contracts/MultiSigWallet.sol#L224

So I think its up to us to decide

File: contracts/contracts/multisig/Multisig.cairo (hunk 9 above)
Comment (taco-paco) [reply contracts/contracts/multisig/Multisig.cairo]: Serum does check for presence of owners in constructor
 https://github.com/project-serum/multisig/blob/d6fbe734dc4026b9c8fdd2f9e78904112150fa74/programs/multisig/src/lib.rs#L40

I thinks that's part of this comment.
https://github.com/smartcontractkit/starknet-multisig/pull/1#discussion_r885548586

Also it looks like that after deployment anyone could execute set_owners with any set of owners. So the first to send such message to Multisig would afterwards control it

File: contracts/contracts/multisig/Multisig.cairo (hunk 2 above)
Comment (taco-paco) [reply contracts/contracts/multisig/Multisig.cairo]: Done

File: contracts/contracts/multisig/Multisig.cairo (hunk 4 above)
Comment (taco-paco) [reply contracts/contracts/multisig/Multisig.cairo]: Done

File: contracts/contracts/multisig/Multisig.cairo (hunk 5 above)
Comment (taco-paco) [reply contracts/contracts/multisig/Multisig.cairo]: Changed the name, added description for `require_tx_valid`

File: contracts/contracts/multisig/Multisig.cairo [hunk 12]
```diff
@@ -326,52 +321,45 @@ end
...
-        range_check_ptr
-    }(
-        to : felt,
-        function_selector : felt,
-        calldata_len : felt,
-        calldata : felt*,
-    ):
+func submit_transaction{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    to : felt, function_selector : felt, calldata_len : felt, calldata : felt*
+):
```
Comment (RodrigoAD) [inline contracts/contracts/multisig/Multisig.cairo]: @epShardlabs  Would it be possible to return the `tx_index`? We'd save some time avoiding looking for the event to get it
```suggestion
) -> (tx_index : felt):
```

File: contracts/contracts/multisig/Multisig.cairo [hunk 13]
```diff
@@ -474,24 +449,104 @@ func execute_transaction{
...
+    alloc_locals
+
+    require_multisig()
+
+    with_attr error_message("invalid num of owners"):
+        assert_not_zero(owners_len)
+    end
+
+    with_attr error_message("owners has to be unique"):
```
Comment (krebernisak) [inline contracts/contracts/multisig/Multisig.cairo]: The other time this is checked on L334, the error message will be "owners not unique".

Let's try to extract common logic and reuse it so we avoid inconsistencies like this.

File: contracts/contracts/multisig/util.cairo [hunk 14]
```diff
@@ -1,6 +1,46 @@
...
+func assert_unique_addresses_impl{range_check_ptr, dict_ptr : DictAccess*}(
+    index : felt, data_len : felt, data : felt*
+):
+    alloc_locals
+    if index == data_len:
+        return ()
+    end
+
+    local el = data[0]
+    assert_not_zero(el)
```
Comment (krebernisak) [inline contracts/contracts/multisig/util.cairo]: This will fail, and then the error message "owners not unique" will be returned.

I suggest we remove this altogether and allow 0x0 addresses so we could burn multisig ownership by `set_owners_and_threshold(owners=[0x0], threshold=1).

Looks like Serum doesn't check for 0x0 owners. cc @archseer

File: contracts/contracts/multisig/Multisig.cairo [hunk 15]
```diff
@@ -148,41 +155,53 @@ func require_confirmed{
...
+    end
+
+    return ()
+end
+
+func require_multisig{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}():
+    let (caller) = get_caller_address()
+    let (contract_address) = get_contract_address()
+
+    with_attr error_message("caller shall be multisig"):
```
Comment (krebernisak) [inline contracts/contracts/multisig/Multisig.cairo]:
```suggestion
    with_attr error_message("caller is not multisig"):
```

File: contracts/contracts/multisig/Multisig.cairo (hunk 3 above)
Comment (krebernisak) [reply contracts/contracts/multisig/Multisig.cairo]: Let's do it like Serum does it:

```rust
        assert_unique_owners(&owners)?;
        require!(!owners.is_empty(), InvalidOwnersLen);

        let multisig = &mut ctx.accounts.multisig;

        if (owners.len() as u64) < multisig.threshold {
            multisig.threshold = owners.len() as u64;
        }
```

tldr; Owners can not be empty, and threshold size is bound to owners size.

File: contracts/contracts/multisig/Multisig.cairo [hunk 16]
```diff
@@ -383,23 +371,21 @@ func submit_transaction{
...
-        range_check_ptr
-    }(tx_index : felt):
+func confirm_transaction{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    tx_index : felt
+):
     require_owner()
     require_tx_exists(tx_index=tx_index)
+    require_tx_valid(tx_index=tx_index)
     require_not_executed(tx_index=tx_index)
+    # TODO: clarify if needed
```
Comment (krebernisak) [inline contracts/contracts/multisig/Multisig.cairo]: This TODO can be removed, right?

File: contracts/contracts/multisig/Multisig.cairo (hunk 7 above)
Comment (krebernisak) [reply contracts/contracts/multisig/Multisig.cairo]: `require_multisig` is ok, but that actually means M/N quorum as the call couldn't be executed any other way from the multisig :)

Yeah, let's do it like this:

- keep proposals closed (only owners)
- execution open (anyone, if quorum reached) 👍

File: contracts/contracts/multisig/Multisig.cairo (hunk 9 above)
Comment (krebernisak) [reply contracts/contracts/multisig/Multisig.cairo]: My only suggestion here is that these two flows:

- set_owners via constructor
- set_owners via public fn

..have different and inconsistent checks.

Which checks exactly we discussed in that other comment: [#1 (comment)](https://github.com/smartcontractkit/starknet-multisig/pull/1/files#:~:text=of%20this%20comment.-,%231%20(comment),-Also%20it%20looks)

But then please make those checks consistent for both flows. For example, constructor can just call `set_owners` internally and make sure it follows the same flow.

File: contracts/contracts/multisig/Multisig.cairo (hunk 3 above)
Comment (taco-paco) [reply contracts/contracts/multisig/Multisig.cairo]: In constructor they use such logic:
```
        assert_unique_owners(&owners)?;
        require!(
            threshold > 0 && threshold <= owners.len() as u64,
            InvalidThreshold
        );
        require!(!owners.is_empty(), InvalidOwnersLen);
```

which is what is done here. I think constructor shalln't be as forgiving as set_owners if invalid threshold was passed then we panic

File: contracts/contracts/multisig/Multisig.cairo (hunk 16 above)
Comment (taco-paco) [reply contracts/contracts/multisig/Multisig.cairo]: Yes, but then we panic if transaction was already confirmed by owner

File: contracts/contracts/multisig/Multisig.cairo (hunk 9 above)
Comment (taco-paco) [reply contracts/contracts/multisig/Multisig.cairo]: That wasn't done because of `require_multisig` check that will fail and I'm not sure that `call_contract` will work since it's not constructed yet. Also I'm not sure that there is problem with inconsistency we have `constructor` flow and `set_owners` flows

File: contracts/contracts/multisig/Multisig.cairo (hunk 13 above)
Comment (taco-paco) [reply contracts/contracts/multisig/Multisig.cairo]: My bad. Will create `require_unique_owners` fn

File: contracts/contracts/multisig/util.cairo (hunk 14 above)
Comment (taco-paco) [reply contracts/contracts/multisig/util.cairo]: Just to give an idea behind what is written in that function:
We have dictionary where all `dict[key]` are set to `default_value`. 
`default_dict_new()`: `Returns a new dictionary where all keys are initialized with a given default value.`
Then for each key we update it's previous value(`default_value`) to `owners[i]`. If in `dict_update` incorrect `prev_value`  was passed then cairo will panic. So if `dict[key]` was updated at least once we will panic on first repeating value(since `prev_value` is not `default_value` anymore).

Check on L22 is necessary because otherwise such array will be possible `[default_value, default_value, ....]`
`dict_new` wasn't used since its not present on StarkNet bacause it requires hints for initialization.

So we can whether create new magic `default_value` like `-1` or use different approach for detecting repeating address(`usort` builtin sort doesn't work).

PS. to future myself not to forget substitute
`assert_not_zero(el - default_value)`

File: contracts/contracts/multisig/Multisig.cairo (hunk 12 above)
Comment (taco-paco) [reply contracts/contracts/multisig/Multisig.cairo]: Will do

File: contracts/contracts/multisig/Multisig.cairo (hunk 3 above)
Comment (archseer) [reply contracts/contracts/multisig/Multisig.cairo]: Let's match this in the constructor then. Calling `set_owners` with 0 owners can make sense if you're trying to burn the multisig, but constructing a multisig with 0 owners is almost 100% a bug because the contract is useless on deploy.

So: can't construct with 0 owners, but can `set_owners` to none.

File: contracts/contracts/multisig/Multisig.cairo (hunk 16 above)
Comment (archseer) [reply contracts/contracts/multisig/Multisig.cairo]: I think @krebernisak means keep the code but remove the `TODO` comment only

File: contracts/contracts/multisig/util.cairo (hunk 14 above)
Comment (archseer) [reply contracts/contracts/multisig/util.cairo]: In the aggregator I've used `key=address` `value=owners_index+1`, using 1-indexed values for `_owners` instead of 0-indexed. That way `0` means undefined and addresses are stored with indexes `1..N`

File: contracts/contracts/multisig/Multisig.cairo (hunk 7 above)
Comment (taco-paco) [reply contracts/contracts/multisig/Multisig.cairo]: Done

File: contracts/contracts/multisig/Multisig.cairo (hunk 8 above)
Comment (taco-paco) [reply contracts/contracts/multisig/Multisig.cairo]: Done.

File: contracts/contracts/multisig/Multisig.cairo (hunk 10 above)
Comment (taco-paco) [reply contracts/contracts/multisig/Multisig.cairo]: Done

File: contracts/contracts/multisig/Multisig.cairo (hunk 12 above)
Comment (taco-paco) [reply contracts/contracts/multisig/Multisig.cairo]: Done

File: contracts/contracts/multisig/util.cairo (hunk 14 above)
Comment (taco-paco) [reply contracts/contracts/multisig/util.cairo]: Used `array.length` as `default_value` to avoid mess with `1..N` indexes

File: contracts/contracts/multisig/util.cairo [hunk 17]
```diff
@@ -1,6 +1,44 @@
...
+
+    return assert_unique_elements_impl(index + 1, data_len, data + 1)
+end
+
+func assert_unique_elements{range_check_ptr}(data_len : felt, data : felt*):
+    alloc_locals
+
+    # default_value shall be not less than data_len
+    # otherwise data[default_value] can have duplicates for index in (default_value, data_len)
+    let (local my_dict_start) = default_dict_new(default_value=data_len)
```
Comment (archseer) [inline contracts/contracts/multisig/util.cairo]: Okay, this is a nice way of handling it. If I understand the default value is `len`, then when calling `dict_update` it will fail setting it if `prev_value` != `len` since that means a value was already inserted from that key?

File: contracts/contracts/multisig/util.cairo [hunk 18]
```diff
@@ -1,6 +1,44 @@
...
+    let my_dict_end = my_dict_start
+
+    assert_unique_elements_impl{range_check_ptr=range_check_ptr, dict_ptr=my_dict_end}(
+        index=0, data_len=data_len, data=data
+    )
+
+    let (finalized_dict_start, finalized_dict_end) = default_dict_finalize(
+        my_dict_start, my_dict_end, data_len
+    )
```
Comment (archseer) [inline contracts/contracts/multisig/util.cairo]: What does this do? Do we even need to run it since we never use the final version of the dict?

File: contracts/contracts/multisig/util.cairo (hunk 17 above)
Comment (taco-paco) [reply contracts/contracts/multisig/util.cairo]: Yeap

File: contracts/contracts/multisig/util.cairo (hunk 18 above)
Comment (taco-paco) [reply contracts/contracts/multisig/util.cairo]: `dict_update`, `dict_write` requires calling `default_dict_finalize` afterwards
`Note that one must eventually call default_dict_finalize() to verify both the internal consistency of the DictAccess entries forming the dictionary and of the consistency with the default value.`

Though I can remove ` let (finalized_dict_start, finalized_dict_end) =` part

File: contracts/contracts/multisig/Multisig.cairo (hunk 12 above)
Comment (taco-paco) [reply contracts/contracts/multisig/Multisig.cairo]: UPD: that actually won't work. Right now there's no 'decent' way to get return value after `invoke`ing the function. The return value of `invoke` is a tx hash & `call` doesn't apply changed state. So the best way here would be to get event's value using tx hash

Comment (krebernisak) [reply contracts/contracts/multisig/Multisig.cairo]: Ah yes, this should not be possible.

Also this GH Issue pending upstream: https://github.com/eqlabs/starknet-multisig/issues/15

Maybe we could actually implement this so Rodrigo is unblocked 🤔

Comment (taco-paco) [reply contracts/contracts/multisig/Multisig.cairo]: This shall be cheap to implement. I would like to create a separate branch for this and then make PR in updatableOwners(or main at that time). What do you think?

Comment (RodrigoAD) [reply contracts/contracts/multisig/Multisig.cairo]: I haven't thought on getting the event with the tx hash, I think that is simple enough

Comment (krebernisak) [reply contracts/contracts/multisig/Multisig.cairo]: @RodrigoAD yeah currently, but we plan to change this - from the linked GH Issue:

> Force the nonce to be subsequent, but not autogenerated. The external entity should generate the next nonce and present it for the contract. The contract checks the previously submitted nonce and makes sure this newest nonce is previous + 1.

@epShardlabs yeah, let's do it in the next PR. I'll sync with Lauri that we can tackle this one 👍

Comment (microbecode) [reply contracts/contracts/multisig/Multisig.cairo]: Kristijan poked me about this. I appreciate if you can tackle this issue as well.
Added some notes to the [issue 15.](https://github.com/eqlabs/starknet-multisig/issues/15)

Comment (taco-paco) [reply contracts/contracts/multisig/Multisig.cairo]: 😉

**Review by archseer — APPROVED** [review]

**Comment by microbecode** [comment]
Any updates on this?
I would like to continue work on the original repo (eqlabs), but that would conflict with this PR and PR#2. Would be great to have these PR'd and merged there first.

**Comment by taco-paco** [comment]
> Any updates on this?
> I would like to continue work on the original repo (eqlabs), but that would conflict with this PR and PR#2. Would be great to have these PR'd and merged there first.

We think changes are ready for an upstream PR. You can find it [here](https://github.com/eqlabs/starknet-multisig/pull/22)
